### PR TITLE
Prevent associated type witnesses from referring to Objective-C type parameters.

### PR DESCRIPTION
**Pull request by DougGregor**
Objective-C type parameters are not recoverable from an instance of
a generic Objective-C class, so we cannot refer to them anywhere in
type metadata. However, they could slip in via associated type
inference, which bypassed the existing checks. This would always crash
later in IRGen; ban it in the type checker so that doesn't happen.

Fixes rdar://problem/34979938.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: lib/Sema/TypeCheckProtocol.cpp [hunk 1]
```diff
@@ -1849,6 +1849,41 @@ void ConformanceChecker::recordTypeWitness(AssociatedTypeDecl *assocType,
 
   assert(!type->hasArchetype() && "Got a contextual type here?");
 
+  // Objective-C's type-erased generics don't allow the
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: Incomplete comment

File: lib/Sema/TypeCheckProtocol.cpp [hunk 2]
```diff
@@ -1849,6 +1849,41 @@ void ConformanceChecker::recordTypeWitness(AssociatedTypeDecl *assocType,
...
+  auto &ctx = assocType->getASTContext();
+  if (ctx.LangOpts.EnableObjCInterop && !type->hasError()) {
+    if (auto classDecl = Adoptee->getClassOrBoundGenericClass()) {
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: This method is already very long, can you factor this out into a new method?

**Comment by DougGregor** [comment]
@slavapestov cleaned up my mess, thanks

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Pull request merged**